admin/src/multitree_bench/mod.rs

```diff
@@ -269,257 +281,443 @@ impl ChainGenerator {
-							}
-						}
-						if path_node.is_some() {
-							break
-						}
-					}
+			// Allocate additions to addition_slots
+			let num_additions = if num_additions >= num_addition_slots {
+				let num_per_slot = num_additions / num_addition_slots;
+				for i in 0..num_addition_slots {
+					addition_slots[i as usize] += num_per_slot;
+				}
+				num_additions - (num_addition_slots * num_per_slot)
+			} else {
+				num_additions
+			};
+			if num_additions > 0 {
+				assert!(num_additions < num_addition_slots);
+				let mut index = rng.next_u32() % num_addition_slots;
+				for _ in 0..num_additions {
+					addition_slots[index as usize] += 1;
+					index = (index + 1) % num_addition_slots;
+				}
+			}
 
-					NUM_PATHS.fetch_add(1, Ordering::SeqCst);
-					match path_node {
-						Some(node) => {
-							NUM_PATHS_SUCCESS.fetch_add(1, Ordering::SeqCst);
-							children.push(node);
-						},
-						None => {
-							// Unable to generate a path so just create a direct node. Use path_rng
-							// to ensure deterministic generation when using only_direct_children.
-							// TODO: This is not returned when only_direct_children is true even
-							// though it is a direct node. Is this ok? only_direct_children is only
-							// used when generating a path so it is ok if it doesn't see these
-							// nodes.
-							children.push(NodeSpec::Direct(
-								tree_index,
-								depth + 1,
-								path_rng.next_u64(),
-							));
-						},
-					}
+			let mut slot_index = 0;
+			for child in child_changes.iter_mut() {
+				if !child.1 {
+					child.3 += addition_slots[slot_index];
+					slot_index += 1;
 				}
 			}
+
+			for _ in slot_index..num_addition_slots as usize {
+				if addition_slots[slot_index] > 0 {
+					// Insert at random index in child_changes
+					let index = rng.next_u32() as usize % (child_changes.len() + 1);
+					child_changes.insert(index, (None, false, 0, addition_slots[slot_index]));
+				}
+				slot_index += 1;
+			}
 		}
 
-		if only_direct_children {
-			return (Vec::new(), children)
+		// Immediate additions
+		for _ in 0..num_immediate_additions {
+			child_changes.push((None, true, 0, 0));
 		}
 
-		// Polkadot doesn't store actual values in branch nodes, only in leaf nodes. Hence the value
-		// stored in the db will only be for the nibble path.
-		let mut size = 4;
-		if num_children == 0 {
-			// Leaf node, so simulate an actual value using the size histogram.
-			size = self.value_length_histogram.sample(rng.next_u64()) as usize;
-			if FORCE_NO_MULTIPART_VALUES {
-				size = std::cmp::min(size, 32760 - 64);
+		// Apply child_changes
+		let mut children = Vec::with_capacity(child_changes.len());
+		for (existing_index, leaf, removals, additions) in child_changes.iter() {
+			if let Some(existing_index) = existing_index {
+				let previous_child_node = &previous_node.unwrap().children[*existing_index];
+				let previous_child = previous_child_node.get_child_as_ref();
+				if *removals > 0 || *additions > 0 {
```

Comment:
I am not too sure about the logic of these removals/additions counters.
The way it should work is : on changed node:
- if removed (eg replacement but there can be removal with no new node in some trie related corner case) then only do Rc minus one, if rc is < 0 then do rc minus one on all children.
- if new add new node.
- if updated, do removed as in first line, and for all children do rc + 1.

The logic I want to assert here is that we got for every changed node this rc + 1 of every of its children and for every pruned node we do the rc - 1 of every of its children.

(seen some io issue running this on a binary trie, if we end up running in similar io issue there is an alternate design, but no use looking at it unless we see such issue)

Reply:
Those values aren’t refcounts, they’re targets for how many values (basically leaf nodes) to add and remove.
This is in the part where it changes the tree for each new commit. It starts with a target number of additions and removals to make for the whole tree then distributes them through child nodes as best it can. Sometimes it adds/removes leaf nodes otherwise it passes the numbers down to the children for them to deal with. They also aren’t guaranteed to be able to make the changes but on average each commit should end up with an appropriate number of changes.
The actual node refcounting happens within the Db and it’ll deal with whether nodes need to be removed.

Reply:
ok, I was wondering abot the number of `NodeRef::Existing(NodeAddress)` in the commit set. I feel like it may have been underestimated (not directly related with this pr though).